src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,200 @@
+
+  // create empty position list for reference segments
+  auto pos_list = std::make_shared<PosList>();
+
+  // check if input table contains reference segments
+  auto& probe_chunk = input_table->get_chunk(ChunkID{0});
+  const auto& probe_segment = probe_chunk.get_segment(ColumnID{0});
+  auto probe_ref_segment = std::dynamic_pointer_cast<ReferenceSegment>(probe_segment);
+  if (probe_ref_segment) {
+    // ref segment
+
+    // set input table to referenced table
+    input_table = probe_ref_segment->referenced_table();
+
+    for (auto position : *(probe_ref_segment->pos_list())) {
+      auto chunk_id = position.chunk_id;
+      auto chunk_offset = position.chunk_offset;
+
+      // get segment for scan operation
+      auto segment = input_table->get_chunk(chunk_id).get_segment(_column_id);
+
+      resolve_data_type(col_type, [&](const auto data_type_t) {
+        using Type = typename decltype(data_type_t)::type;
+
+        // assert that type of scan value and type of segment (column) are the same
+        Assert((_search_value.type() == typeid(Type)), "Type of search value and column should be the same");
+
+        // cast to value segment to check whether we have value segment or dictionary segment
+        // if cast returns null ptr we have dictionary segment
+        // if successful cast we have value segment
+        auto value_segment = std::dynamic_pointer_cast<ValueSegment<Type>>(segment);
+
+        // get corresponding comparator method for ScanType
+        auto _comparator = _get_comparator_method<Type>();
+
+        if (value_segment) {
+          // value Segment
+
+          // get actual value for position list entry
+          auto comparison_value = type_cast<Type>(value_segment->values()[chunk_offset]);
+
+          // compare to search value
+          if (_comparator(type_cast<Type>(_search_value), comparison_value)) {
+            pos_list->push_back(RowID{chunk_id, chunk_offset});
+          }
+        } else {
+          // dictionary Segment
+
+          // cast to dictionary segment
+          auto dictionary_segment = std::dynamic_pointer_cast<DictionarySegment<Type>>(segment);
+
+          // get actual value for position list entry
+          auto comparison_value = type_cast<Type>(dictionary_segment->get(chunk_offset));
+
+          // compare to search value
+          if (_comparator(type_cast<Type>(_search_value), comparison_value)) {
+            pos_list->push_back(RowID{chunk_id, chunk_offset});
+          }
+        }
+      });
+    }
+  } else {
+    // no ref segment
+    auto chunk_count = input_table->chunk_count();
+
+    // iterate over chunks
+    for (auto chunk_id = ChunkID{0}; chunk_id < chunk_count; ++chunk_id) {
+      auto& chunk = input_table->get_chunk(chunk_id);
+      std::shared_ptr<BaseSegment> segment = chunk.get_segment(_column_id);
+
+      // Fill position list for different segment types
+      resolve_data_type(col_type, [&](const auto data_type_t) {
```

Comment:
See above. Why should we do that for every row if we don't have reference segments?